url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/utils/StatisticalSummary.java

```diff
@@ -0,0 +1,327 @@
+      s1.mean(), s2.mean(), s1.sumsq, s2.sumsq, s1.n, s2.n, true);
+  }
+*/
+
+    public static double sigDiff(StatisticalSummary s1, StatisticalSummary s2) {
+        return StatisticalTests.tNotPaired(
+                s1.mean(), s2.mean(), s1.sumSquareDiff(), s2.sumSquareDiff(), s1.n, s2.n, true);
+    }
+
+    /**
+     * returns the sum of the squares of the differences
+     * between the mean and the ith values
+     */
+    public double sumSquareDiff() {
+        return sumsq - n * mean() * mean();
+    }
+
+    private void computeStats() {
+        if (!valid) {
+            mean = sum / n;
+            gm = Math.exp(logsum / n);
+            double num = sumsq - (n * mean * mean);
+            if (num < 0) {
+                // avoids tiny negative numbers possible through imprecision
+                num = 0;
+            }
+            // System.out.println("Num = " + num);
+            sd = Math.sqrt(num / (n - 1));
+            // System.out.println(" Test: sd = " + sd);
+            // System.out.println(" Test: n = " + n);
+            valid = true;
+        }
+    }
+
+    public double sd() {
+        if (!valid)
+            computeStats();
+        return sd;
+    }
+
+    public int n() {
+        return n;
+    }
+
+    public double stdErr() {
+        return sd() / Math.sqrt(n);
+    }
+
+    public void add(StatisticalSummary ss) {
+        // implications for Watch?
+        n += ss.n;
+        sum += ss.sum;
+        sumsq += ss.sumsq;
+        logsum += ss.logsum;
+        max = Math.max(max, ss.max);
+        min = Math.min(min, ss.min);
+        valid = false;
+    }
+
+    public void add(double d) {
+        n++;
+        sum += d;
+        sumsq += d * d;
+        if (d > 0) {
+            logsum += Math.log(d);
+        }
+        min = Math.min(min, d);
+        max = Math.max(max, d);
+        if (watch != null) {
+            watch.note(d);
+        }
+        valid = false;
+    }
+
+    public void add(Number n) {
+        add(n.doubleValue());
+    }
+
+    public void add(double[] d) {
+        for (int i = 0; i < d.length; i++) {
+            add(d[i]);
+        }
+    }
+
+    public void add(Vector v) {
+        for (int i = 0; i < v.size(); i++) {
+            try {
+                add(((Number) v.elementAt(i)).doubleValue());
+            } catch (Exception e) {
+            }
+        }
+    }
+
+    public String toString() {
+        String s = (name == null) ? "" : name + "\n";
+        s += " min = " + min() + "\n" +
+                " max = " + max() + "\n" +
+                " ave = " + mean() + "\n" +
+                " sd  = " + sd() + "\n" +
+               // " se  = " + stdErr() + "\n" +
+               // " sum  = " + sum + "\n" +
+               // " sumsq  = " + sumsq + "\n" +
+               // " watch = " + watch + "\n" +
+                " n   = " + n;
+        return s;
+
+    }
+
+    public void save(String path) {
+        try {
+            ObjectOutputStream oos =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *oos*. The resource is referenced at lines 244, 246. The resource is closed at line 247. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *oos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).